### ISSUE TITLE
Exceptions that are uncaught result in a horrible debugging experience

### ISSUE THREAD

**Issue by sc2ad** [sc2ad/beatsaber-hook#128]
What do I mean? Well, if we have an `RunMethodRethrow` call, and the exception thrown in the C# method (which then bubbles to a C++ method) is uncaught, the exception will travel all the way to `terminate`, but with a completely unwound stack.

This makes it virtually impossible to tell _where_ the damn exception was actually thrown, thus making this form of throwing horrible from a debugging perspective-- all uncaught exceptions will bubble to the caller's handler, which could be `libcodegen`, `libil2cpp`, or even `libmod` (perhaps even ANOTHER mod, depending on who called who). Point is, it's horrible and tells us LESS than if we just caused a forced crash.

So, how do we fix this? I see a few high level fixes (for common cases, we will do some more analysis of some stuff in a moment)
1. Don't throw anymore, lets just crash again (maybe more optionally?)
2. Throw exceptions that have stacktrace info, and later log it/keep track of it (maybe on exception dtor or something) This is actually pretty hard because we need to somehow handle cases where uncaught exceptions bubble across domains (ex: exception thrown bubbles to `libil2cpp` as opposed to bubbling to `libcodegen`)
3. Throw exceptions that are just instantly caught and handle them somehow

All of these have some pretty glaring issues--
1. The first approach sounds great for most cases (especially having crashing be optional, perhaps allowing for UB), but when you realize that you want to be able to catch C# exceptions, things become much trickier-- ALL WHILE still allowing them to be caught in an exceptional case (completely uncaught)
2. This sounds like it solves some issues, we can keep track of exceptions with stack trace info and maybe even display that info! Sounds great on paper, and realistically would work for MOST cases (when exceptions bubble up to a library we are in charge of), but outside of that things become just as bad. This is actually not the worst possible solution, but grabbing relevant stack trace info is not actually truly trivial-- getting our calling function isn't TOO bad, but functions above that become much more obnoxious, and doing all of this just to simply catch the exception seems like a bit of a waste in the C#-caught case described earlier.
3. This also sounds great, but there are some tricky issues to deal with here as well. Namely, we need to add additional boilerplate for exceptions that bubble up that would be ABOUT to cross a boundary, but then we run into a similar issue of losing stack trace info, so we would need these exceptions to hold SOME information about their stacktrace so that we could at least log that in our overall catches before we would otherwise throw to `libil2cpp` which would (probably inevitably) crash.

Soooooooo, yeah. Tough problems, tough solutions. This is also relevant with #125 which has some information-- also errs on the side of just logging useful debugging info all the time, which I'd like to avoid if possible.

I feel like there's probably a clever approach here that could take advantage of multiple of these approaches, or perhaps a new one altogether (maybe we implement our own pseudo-throw system? One where if the exception would bubble, we handle it accordingly)

Or we could maybe take advantage of hooking something like `__cxa_throw` to set our stacktrace somehow, and then use that if we need to except, but that still runs the same problems as before, where we need the stacktrace info.

Ultimately, open to ideas and feedback, hopefully this gives a bit of understanding to those curious and sparks some discussion around interesting thoughts that might be possible.

**Comment by sc2ad** [comment]
So, I did a little bit of work on this front today. The main solution that I worked towards was solution 2. The downsides of this implementation are many-- even throwing an instantly caught C# exception causes us to allocate a GIGANTIC amount of stack trace space. We should probably tune this or perhaps have it be completely instance independent, but ultimately that is a big issue I would like to resolve.

In addition, due to x86-64 differences, addresses are a bit trickier to read, harder to load via symbol lookups, etc. so I manually added comments to prove that it works as intended.

Finally, this would also require additional boilerplate to the hook and custom type macros in order to catch these exceptions and handle them appropriately before they get forwarded to `libil2cpp.so` which would be less than ideal, debugging-wise.
https://godbolt.org/z/3WhhMrsd8

**Comment by Lauriethefish** [comment]
Is there any chance we can get an option to disable the logging of `RunMethodException`? (i.e. if passing the exception to il2cpp is what we want in the first place)

**Comment by sc2ad** [comment]
It should be noted that the exception is only logged verbosely if it was uncaught in all other contexts and bubbled all the way up to the point where it WOULD be thrown across a library boundary and result in UB. As such, my recommendation for if you wish to forward an exception EXPLICITLY to il2cpp is to add a try catch over the critical potion yourself, that way you are explicit and very clear about what should happen. The alternative case, where the exception remains uncaught, should certainly be handled differently in practice, and due to how logging is already present in many bs-hook cases, this being another is not a bad design choice, in my opinion.

That being said, there is certainly room for argument with the gigantic exceptions being allocated, though you can also argue that an exception is, well, exceptional, so normal behavior wouldn't need to allocate with such a large amount of space.

I'd say the main unfortunate issues are the fact that exceptions (even caught ones) have to have the stack trace info, which now makes handling a standard c# exception a bit slower, since the stack trace is determined and then subsequently ignored. But this is actually really hard to fix in practice, as we don't have a good way of knowing if an exception is to be caught, or what types should be caught, and how.

I have considered making a new type of run method call that would allow you specific which types of exceptions you would like to handle, and perhaps support specific handling logic pass-by-arg, as opposed to simply returning the caught exception, but fundamentally even if this existed, it would not be used in codegen. If it was used in codegen, the template parameters would have to be exposed to the most relevant dev, which would result in it all being in headers, so codegen would be entirely templated and header only, which is a pain I will not introduce as of now.

Adding some sort of global catch handler that varies is defacto supported-- for example, you could ignore rethrowing the exceptions into the il2cpp domain and instead opt to simply safe abort on any exception. You could do this by simply creating your own hook structure, akin to how the macros do so, but replacing the hook method pointer with a pointer to your own wrapper method that performs whatever logic you prefer.

Tldr, yeah, you can do it yourself, or you can choose to ignore it, or you can change things up, but I won't change the current implementation I think.

**Comment by sc2ad** [comment]
Okay, this has been resolved in v3.7.4+, using the implementation that I described earlier in this issue.
